File: app/src/sim/reconcile.ts
// The reconcile loop (spec §9): observe → derive State → converge gaps within
// the approved envelope. Converge actions ONLY — it never mutates desired state
// (that is the Author class).

import type { Manifest, ResourceID, Health } from "./model";
import type { Provider } from "./provider";
import { derive, type Control, type State } from "./state";

export type DriftPolicy = "enforce" | "warn" | "ignore";
export type Action = "none" | "apply" | "delete" | "hold" | "warn";

export interface Options {
  driftPolicy?: DriftPolicy;
  stalenessBudgetMs?: number;
  /** Consecutive failed self-heals before the breaker trips to Stalled (§9). */
  flapThreshold?: number;
}

export interface Status {
  id: ResourceID;
  state: State;
  health: Health;
  action: Action;
  reason: string;
}

export class Reconciler {
  private frozen = new Set<ResourceID>();
  private stalled = new Set<ResourceID>();
  private attempts = new Map<ResourceID, number>();
  private driftPolicy: DriftPolicy;
  private stalenessBudgetMs: number;
  private flapThreshold: number;

  constructor(
    private p: Provider,
    opts: Options = {},
  ) {
    this.driftPolicy = opts.driftPolicy ?? "enforce";
    this.stalenessBudgetMs = opts.stalenessBudgetMs ?? 0;
    this.flapThreshold = opts.flapThreshold ?? 3;
  }

  /** Break-glass: suspend reconciliation of a resource (spec §7). */
  freeze(id: ResourceID) {
    this.frozen.add(id);
  }
  unfreeze(id: ResourceID) {
    this.frozen.delete(id);
  }
  isFrozen(id: ResourceID) {
    return this.frozen.has(id);
  }

  /** Circuit-breaker state (spec §9 reconciler self-protection). */
  isStalled(id: ResourceID) {
    return this.stalled.has(id);
  }
  /** Resolve an incident: clear the breaker so reconciliation resumes. */
  resetBreaker(id: ResourceID) {
    this.stalled.delete(id);
    this.attempts.delete(id);
  }

  step(m: Manifest, nowMs: number): Status[] {
    const obs = this.p.observeAll();
    const byId = new Map(obs.map((o) => [o.id, o]));
    const out: Status[] = [];

    for (const [id, d] of Object.entries(m.resources)) {
      const o =
        byId.get(id) ??
        { id, exists: false, spec: {}, health: "Unknown" as Health, appliedGeneration: 0, observedAtMs: 0 };

      // External: observe-only — the reconciler never converges it (§1).
      if (d.lifecycle === "external") {
        const st = derive(d, o, "Settled", nowMs, this.stalenessBudgetMs);
        out.push({
          id,
          state: st,
          health: o.health,
          action: "none",
          reason: st === "Degraded" ? "external dependency degraded — observe-only, cannot remediate" : "external SaaS — observed only",
        });
        continue;
      }

      // Job: run-to-completion. Launch it once; then hold — reaching succeeded
      // is success, not drift, and the sim re-runs it on schedule.
      if (d.lifecycle === "job") {
        const st = derive(d, o, "Settled", nowMs, this.stalenessBudgetMs);
        const sj: Status = { id, state: st, health: o.health, action: "none", reason: "" };
        if (!o.exists) {
          this.p.apply(d);
          sj.action = "apply";
          sj.reason = "launching job";
        } else if (st === "Succeeded") {
          sj.reason = "completed — success, not drift";
        } else if (st === "Running") {
          sj.reason = "run in progress";
        } else if (st === "Failed") {
          sj.reason = "run failed — will retry on schedule";
        } else {
          sj.reason = "scheduled — pending";
        }
        out.push(sj);
        continue;
      }

      const control: Control = this.frozen.has(id)
        ? "Frozen"
        : this.stalled.has(id)
          ? "Stalled"
          : "Settled";
      let st = derive(d, o, control, nowMs, this.stalenessBudgetMs);

      // Circuit breaker (§9): a resource that keeps failing self-heal flaps,
      // then trips to Stalled — the reconciler stops retrying into a crash-loop
      // and escalates to a human.
      if (st === "Degraded") {
        const n = (this.attempts.get(id) ?? 0) + 1;
        this.attempts.set(id, n);
        if (n > this.flapThreshold) {
          this.stalled.add(id);
          st = "Stalled";
        }
      } else if (st === "Converged") {
        this.attempts.delete(id);
        this.stalled.delete(id);
      }

      const s: Status = { id, state: st, health: o.health, action: "none", reason: "" };

      switch (st) {
        case "Converged":
          s.reason = "matches spec and healthy";
          break;
        case "Converging":
          this.p.apply(d);
          s.action = "apply";
          s.reason = "authored change rolling out";
          break;
        case "Degraded":
          this.p.apply(d);
          s.action = "apply";
          s.reason = "self-heal: unhealthy, re-provisioning";
          break;
        case "Drifted":
          if (this.driftPolicy === "enforce") {
            this.p.apply(d);
            s.action = "apply";
            s.reason = "drift: unauthored change, correcting";
          } else if (this.driftPolicy === "warn") {
            s.action = "warn";
            s.reason = "drift detected (policy=warn, not correcting)";
          } else {
            s.reason = "drift ignored (policy=ignore)";
          }
          break;
        case "Unknown":
          s.action = "hold";
          s.reason = "telemetry stale/missing — holding (fail-safe)";
          break;
        case "Frozen":
          s.action = "hold";
          s.reason = "break-glass: reconciliation suspended";
          break;
        case "Stalled":
          s.action = "hold";
          s.reason = "stalled — needs a human";
          break;
      }
      out.push(s);
    }

    // Retire resources absent from desired (unauthored existence) under enforce.
    if (this.driftPolicy === "enforce") {
      for (const o of obs) {
        if (m.resources[o.id] || !o.exists || this.frozen.has(o.id)) continue;
        this.p.delete(o.id);
        out.push({ id: o.id, state: "Drifted", health: o.health, action: "delete", reason: "not in desired state — retiring" });
      }
    }

    out.sort((a, b) => (a.id < b.id ? -1 : 1));
    return out;
  }
}

export function allConverged(statuses: Status[]): boolean {
  return statuses.length > 0 && statuses.every((s) => s.state === "Converged");
}

File: reconcile/reconcile.go
// Package reconcile implements the Trellis reconcile spine (build-sequence step
// 1, spec §9): observe the cloud through the provider port, derive each
// resource's State, and converge gaps within the approved envelope —
// continuously. Self-healing is just this loop running forever instead of once.
//
// The reconciler performs only Converge actions (spec §7). It never mutates
// desired state — that is the Author class, and it happens elsewhere (the
// manifest). This is the one law: desired state changes only through Author;
// everything else converges toward it.
package reconcile

import (
	"context"
	"sort"
	"time"

	"github.com/hakkiki/trellis/model"
	"github.com/hakkiki/trellis/provider"
	"github.com/hakkiki/trellis/state"
)

// DriftPolicy controls how unauthored divergence is handled, per scope (spec
// §4). It defaults to Enforce.
type DriftPolicy int

const (
	// Enforce auto-corrects drift back to desired.
	Enforce DriftPolicy = iota
	// Warn reports drift but does not correct it.
	Warn
	// Ignore leaves drift alone (e.g. a resource hand-managed during migration).
	Ignore
)

// Options configure a Reconciler.
type Options struct {
	// DriftPolicy decides what to do with unauthored divergence.
	DriftPolicy DriftPolicy
	// StalenessBudget bounds how old an observation may be before the resource
	// is treated as Unknown and held (Invariant 7). Zero disables the check.
	StalenessBudget time.Duration
}

// Action names what the reconciler decided to do about a resource in a Step.
type Action int

const (
	// ActNone means the resource is converged or intentionally held.
	ActNone Action = iota
	// ActApply means the reconciler converged the resource toward desired.
	ActApply
	// ActDelete means the reconciler retired a resource absent from desired.
	ActDelete
	// ActHold means the reconciler deliberately did nothing (Unknown/Frozen).
	ActHold
	// ActWarn means drift was detected but policy says don't correct it.
	ActWarn
)

func (a Action) String() string {
	switch a {
	case ActApply:
		return "apply"
	case ActDelete:
		return "delete"
	case ActHold:
		return "hold"
	case ActWarn:
		return "warn"
	default:
		return "none"
	}
}

// Status is the reconciler's per-resource report from one Step.
type Status struct {
	ID     model.ResourceID
	State  state.State
	Health model.Health
	Action Action
	// Reason is a short, human-readable justification — the seed of the proof
	// discipline ("no magic": every action is explainable).
	Reason string
}

// Reconciler converges a provider toward desired state.
type Reconciler struct {
	p      provider.Provider
	opts   Options
	frozen map[model.ResourceID]bool // break-glass: reconciliation suspended
}

// New creates a Reconciler over a provider.
func New(p provider.Provider, opts Options) *Reconciler {
	return &Reconciler{p: p, opts: opts, frozen: map[model.ResourceID]bool{}}
}

// Freeze suspends reconciliation of a resource — the effect of break-glass (spec
// §7). The resource enters Frozen and the reconciler will not touch it (the debt
// is repaid out of band, by an Author action).
func (r *Reconciler) Freeze(id model.ResourceID)   { r.frozen[id] = true }
func (r *Reconciler) Unfreeze(id model.ResourceID) { delete(r.frozen, id) }

// Step runs one reconcile pass against the manifest and returns the derived
// state and action for every resource considered. now is the reconciler's time
// source (use the provider's clock so freshness checks agree).
func (r *Reconciler) Step(ctx context.Context, m model.Manifest, now time.Time) ([]Status, error) {
	obs, err := r.p.ObserveAll(ctx)
	if err != nil {
		return nil, err
	}
	byID := make(map[model.ResourceID]model.Observation, len(obs))
	for _, o := range obs {
		byID[o.ID] = o
	}

	var out []Status

	// Converge every desired resource.
	for id, d := range m.Resources {
		o := byID[id] // zero value has Exists == false, which is correct
		ctrl := state.ControlSettled
		if r.frozen[id] {
			ctrl = state.ControlFrozen
		}
		st := state.Derive(d, o, ctrl, now, r.opts.StalenessBudget)
		status := Status{ID: id, State: st, Health: o.Health}

		switch st {
		case state.StateConverged:
			status.Action, status.Reason = ActNone, "matches spec and healthy"
		case state.StateConverging:
			if err := r.p.Apply(ctx, d); err != nil {
				return out, err
			}
			status.Action, status.Reason = ActApply, "authored change rolling out"
		case state.StateDegraded:
			// Self-heal: correct-but-unhealthy → re-provision within envelope.
			if err := r.p.Apply(ctx, d); err != nil {
				return out, err
			}
			status.Action, status.Reason = ActApply, "self-heal: unhealthy, re-provisioning"
		case state.StateDrifted:
			switch r.opts.DriftPolicy {
			case Enforce:
				if err := r.p.Apply(ctx, d); err != nil {
					return out, err
				}
				status.Action, status.Reason = ActApply, "drift: unauthored change, correcting"
			case Warn:
				status.Action, status.Reason = ActWarn, "drift detected (policy=warn, not correcting)"
			default: // Ignore
				status.Action, status.Reason = ActNone, "drift ignored (policy=ignore)"
			}
		case state.StateUnknown:
			// Fail-safe: never act on stale/missing data (Invariant 7).
			status.Action, status.Reason = ActHold, "telemetry stale/missing — holding"
		case state.StateFrozen:
			status.Action, status.Reason = ActHold, "break-glass: reconciliation suspended"
		case state.StateStalled:
			status.Action, status.Reason = ActHold, "stalled — needs a human"
		}
		out = append(out, status)
	}

	// Retire resources that exist but desired state no longer contains — these
	// are unauthored existence (drift), corrected under Enforce.
	if r.opts.DriftPolicy == Enforce {
		for _, o := range obs {
			if _, wanted := m.Resources[o.ID]; wanted || !o.Exists {
				continue
			}
			if r.frozen[o.ID] {
				continue
			}
			if err := r.p.Delete(ctx, o.ID); err != nil {
				return out, err
			}
			out = append(out, Status{
				ID: o.ID, State: state.StateDrifted, Health: o.Health,
				Action: ActDelete, Reason: "not in desired state — retiring",
			})
		}
	}

	sort.Slice(out, func(i, j int) bool { return out[i].ID < out[j].ID })
	return out, nil
}

// Converged reports whether every desired resource is Converged.
func Converged(statuses []Status) bool {
	for _, s := range statuses {
		if s.State != state.StateConverged {
			return false
		}
	}
	return len(statuses) > 0
}
